### PR TITLE
ASUS Strix 2080 TI OC

### PR DESCRIPTION
**Pull request by MarshallAsch**
Closes: #222 

Added the ASUS Strix 2080 Ti OC driver that is currently working for me along with documentation on how to use the device.
- added the ASUS 2080ti rev A. and the EGVA 1080 FTW to the udev rule list

This is the same as #228 that accidentally got closed when the destination branch got deleted.

Steps to do before this can be merged:

- [x] Move driver over to the `nvidia.py` driver
- [x] Create automated tests.

### CODE REVIEW

**Comment by MarshallAsch** [comment]
@jonasmalacofilho This is now ready for review, all of the tasks that I still had to complete should be resolved.

**Comment by jonasmalacofilho** [comment]
@MarshallAsch reviewing...

Note: I'm making some small fixes myself, please don't make new changes for the time being.

**Review by jonasmalacofilho — APPROVED** [review]

**Comment by jonasmalacofilho** [comment]
Please take a look and see if you spot any problems with the changes in 7899412.

---

Apart from that, I also removed the assertions on volatile==true in the test cases (8ee1705).  I'll adjust that accordingly (and, if necessary, reintroduce them) once I better understand when volatile changes persist (see discord).

The corresponding docs have `FIXME` for the same reason.

File: docs/nvidia-guide.md [hunk 1]
```diff
@@ -85,10 +88,10 @@ Experimental. Only RGB lighting supported.
 
 Unsafe features:
 
-- `smbus`: enable SMBus support; SMBus devices may not tolerate writes or reads they do not expect
+- `smbus`: enable SMBus support; SMBus devices may not tolerate writes or reads
+  they do not expect
 - `rog_turing`: enable access to the specific graphics cars
```
Comment (MarshallAsch) [inline docs/nvidia-guide.md:93]: spelling. `cars` -> `cards`.

File: extra/linux/71-liquidctl.rules [hunk 2]
```diff
@@ -17,15 +17,14 @@
...
-# NVIDIA ASUS Strix 2080 Ti OC GPU
-KERNEL=="i2c-[0-9]*", ATTRS{vendor}=="0x10de", ATTRS{subsystem_vendor}=="0x1043", ATTRS{device}=="0x1e07", ATTRS{subsystem_device}=="0x866a", TAG+="uaccess"
-
-# NVIDIA EVGA GTX 1080 FTW
-KERNEL=="i2c-[0-9]*", ATTRS{vendor}=="0x10de", ATTRS{subsystem_vendor}=="0x3842", ATTRS{device}=="0x1b80", ATTRS{subsystem_device}=="0x6286", TAG+="uaccess"
-
 # Asetek 690LC (also EVGA CLC or NZXT Kraken)
 SUBSYSTEMS=="usb", ATTRS{idVendor}=="2433", ATTRS{idProduct}=="b200", TAG+="uaccess"
 
+# ASUS Strix 2080 Ti OC
```
Comment (MarshallAsch) [inline extra/linux/71-liquidctl.rules:23]: Should the usb and the i2c rules be intermixed or would it be better to have sections, for USB and for i2c? To make it easier if someone only wants some rules and not all.

File: liquidctl/driver/nvidia.py [hunk 3]
```diff
@@ -1,7 +1,6 @@
 """liquidctl drivers for NVIDIA graphics cards.
-liquidctl drivers for ASUS Strix RTX 2080 Ti OC graphics card.
 
-Copyright (C) 2020–2020  Jonas Malacoi, Marshall Asch and contributors
+Copyright (C) 2020–2020  Jonas Malaco, Marshall Asch and contributors
```
Comment (MarshallAsch) [inline liquidctl/driver/nvidia.py:3]: oops, vim mistake, it seems I was already in insert mode.

File: liquidctl/driver/nvidia.py [hunk 4]
```diff
@@ -41,24 +51,16 @@ class EvgaPascal(SmbusDriver):
...
-        def __new__(cls, value, required_colors):
-            obj = bytes.__new__(cls, [value])
-            obj._value_ = value
-            obj.required_colors = required_colors
-            return obj
-
+    class Mode(_ModeEnum):
         OFF = (0x00, 0)
         FIXED = (0x01, 1)
         RAINBOW = (0x02, 0)
```
Comment (MarshallAsch) [inline liquidctl/driver/nvidia.py:57]: I noticed for your GPU there is no mode `0x03` or `0x04`, is it possible that those exist and are a fade / flash mode? It seems odd that they just skipped a couple numbers

File: liquidctl/driver/nvidia.py [hunk 5]
```diff
@@ -238,32 +230,34 @@ def probe(cls, smbus, vendor=None, product=None, address=None, match=None,
...
+                        val2 = smbus.read_byte_data(address, 0x21)
+                    except:
+                        pass
+                    smbus.close()
+
+                    if val1 << 8 | val2 == ASUS_GPU_MAGIC_VALUE:
+                        selected_address = address
+                        break
+            else:
+                selected_address = cls._SENTINEL_ADDRESS
```
Comment (MarshallAsch) [inline liquidctl/driver/nvidia.py:252]: What is the reason for `_SENTINEL_ADDRESS`? I get that it is a default, but not sure why it has that value, and does that name have any significance?

File: liquidctl/driver/nvidia.py (hunk 4 above)
Comment (jonasmalacofilho) [reply liquidctl/driver/nvidia.py:57]: I though so too, but there doesn't seem to be nothing there.  Trying to fill in this gap is also why before I had these encoded in bits.

File: liquidctl/driver/nvidia.py (hunk 5 above)
Comment (jonasmalacofilho) [reply liquidctl/driver/nvidia.py:252]: I described it in the commit message: 7899412.

TL; DR. To prevent a card opened without `unsafe` to later be used with `unsafe`, and thus access an address we didn't actually validate.

The sentinel is checked for in `get_status` and `set_color`, and it also has a large value so that even if we forget, the lower level APIs will still probably refuse it.

File: extra/linux/71-liquidctl.rules (hunk 2 above)
Comment (jonasmalacofilho) [reply extra/linux/71-liquidctl.rules:23]: You're right, that would be better.
